src/Tools/ExternalAccess/Razor/Cohost/RazorSemanticTokensRefreshQueueWrapper.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens;
+using Newtonsoft.Json;
+using Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+
+[ExportCohostLspServiceFactory(typeof(IRazorSemanticTokensRefreshQueue)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class RazorSemanticTokensRefreshQueueWrapperFactory() : ILspServiceFactory
+{
+    public ILspService CreateILspService(LspServices lspServices, WellKnownLspServerKinds serverKind)
+    {
+        var semanticTokensRefreshQueue = lspServices.GetRequiredService<SemanticTokensRefreshQueue>();
+
+        return new RazorSemanticTokensRefreshQueueWrapper(semanticTokensRefreshQueue);
+    }
+
+    internal class RazorSemanticTokensRefreshQueueWrapper(SemanticTokensRefreshQueue semanticTokensRefreshQueue) : IRazorSemanticTokensRefreshQueue, IDisposable
+    {
+        public void Initialize(string clientCapabilitiesString)
+        {
+            var clientCapabilities = JsonConvert.DeserializeObject<VSInternalClientCapabilities>(clientCapabilitiesString) ?? new();
+            semanticTokensRefreshQueue.Initialize(clientCapabilities);
```

Comment:
will this result in initialize being called twice? (one for the main on initialized, once for razor oninitialize?

Reply:
Yesn't. This Razor class only initializes in the AlwaysActivatedInProc server, and its main oninitialized will call the queue to initialize, but since it doesn't handle semantic tokens, that call does nothing. That's why I had to extract the initialize method out in the queue itself, so I could call it here separately.

Reply:
If it isn't too hard can we make the Initialize resilient to being called twice (in a potential race).  I worry we won't realize its being called twice (e.g. when we enable cohosting in vscode which does call semantic tokens initialize).

Reply:
Oh, the initialize call in the queue itself is resilient. It doesn't lock though. Do you think that's worth it? Are we likely to fire IOnInitialized handlers in parallel ever?

Reply:
It seems like it could race if two calls to initialize run in parallel (might recreate the actual batching work queue and subscribe twice to the events).

But now that you mention it, if both calls are coming from an IOnInitialized service we do technically run them serially
https://github.com/dotnet/roslyn/blob/4f6a970521eb50624984d754305b02a51ca0ca00/src/Features/LanguageServer/Protocol/Handler/ServerLifetime/InitializedHandler.cs#L29

So it probably wouldn't get called twice as long as that is the case.  Would be good to add comment on this mentioning that is why its safe.

Reply:
Yep, will do. The indirection makes it opaque, but yes the call to this method comes from code in the Razor repo which is called from that said IOnInitialized loop